lib/compiler/src/beam_doc.erl

```diff
@@ -0,0 +1,241 @@
+%%
+%% %CopyrightBegin%
+%%
+%% Copyright Ericsson AB 2023-2028. All Rights Reserved.
+%%
+%% Licensed under the Apache License, Version 2.0 (the "License");
+%% you may not use this file except in compliance with the License.
+%% You may obtain a copy of the License at
+%%
+%%     http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing, software
+%% distributed under the License is distributed on an "AS IS" BASIS,
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%
+%% %CopyrightEnd%
+%%
+%% Purpose : Generate documentation as per EEP-48
+
+
+-module(beam_doc).
+
+-feature(maybe_expr, enable).
+
+-moduledoc "
+Tool to convert an Erlang program in its abstract form to EEP-48 format.
+
+### Example:
+
+```erlang
+1> compile:file(test, [beam_docs]).
+```
+".
+
+-export([main/2]).
+
+-include_lib("kernel/include/eep48.hrl").
+
+-doc "
+Internal record used when transforming Erlang abstract syntax into EEP-48
+documentation format.
+".
+
+-record(docs, {cwd                 :: unicode:chardata(),             % Cwd
+               module_name = undefined  :: unicode:chardata() | undefined,
+               doc    = undefined  :: unicode:chardata(), % Function/type/callback local doc
+               meta   = maps:new() :: map()}).      % Function/type/callback local meta
+
+-doc "
+Internal documentation type for generating EEP-48 documentation format.
+".
+-type internal_docs() :: #docs{}.
+
+
+-doc "
+Transforms an Erlang abstract syntax form into EEP-48 documentation format.
+".
+-spec main(term(), term()) -> term().
+main(Dirname, AST) ->
+    try
+        {ModuleDocAnno, ModuleDoc} = extract_moduledoc(AST),
+        DocFormat = extract_docformat(AST),
+        Docs = extract_documentation(AST, new_state(Dirname)),
+        #docs_v1{
+           format = DocFormat,
+           anno = ModuleDocAnno,
+           module_doc = #{ <<"en">> => ModuleDoc },
+           docs = Docs }
+    catch E:R:ST ->
+            erlang:raise(E, R, ST)
+    end.
+
+-spec extract_docformat(AST :: [tuple()]) -> DocFormat :: binary().
+extract_docformat(AST) ->
+    case lists:keysearch(docformat, 3, AST) of
```

Comment:
The compiler has a convention to import all functions used from the lists module. We should do the same here.